src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1461 @@
+
+        if self_attn_mask is not None:
+            self_attn_mask = self_attn_mask.unsqueeze(0)
+            attn_weights_main = attn_weights_main + self_attn_mask
+
+        attn_weights_main = softmax(
+            attn_weights_main, dim=-1, onnx_trace=self.onnx_trace,
+        ).type_as(attn_weights_main)
+        attn_weights_main = F.dropout(attn_weights_main, p=self.dropout, training=self.training)
+
+        attn_main = torch.bmm(attn_weights_main, v_main)
+        attn_main = attn_main.transpose(0, 1).contiguous().view(1, real_tgt_len, bsz, embed_dim)
+        attn_main = self.out_proj(attn_main)
+
+        # [ngram, B*head, T, c]
+        q_ngram = torch.cat(q_predict_list, 0).view(self.ngram, -1, real_tgt_len, self.head_dim)
+        # [ngram, B*head, 2*T, c]
+        k_ngram = torch.cat([torch.cat([k_main, k_p], 1).unsqueeze(0) for k_p in k_predict_list], 0)
+        # below code slower than above for loop
+        # k_ngram = torch.cat([k_main.unsqueeze(0).repeat(self.ngram, 1, 1, 1) , torch.cat(k_predict_list).view(self.ngram, -1, real_tgt_len, self.head_dim)], 2)
+
+        # [ngram, T, B, C]
+        h_ngram = torch.cat(h_predict_list, 0).view(self.ngram, real_tgt_len, bsz, embed_dim)
+
+        # [ngram, B*head, 2*T, c]
+        v_ngram = torch.cat([torch.cat([v_main, v_p], 1).unsqueeze(0) for v_p in v_predict_list], 0)
+        # below code slower than above for loop
+        # v_ngram = torch.cat([v_main.unsqueeze(0).repeat(self.ngram, 1, 1, 1) , torch.cat(v_predict_list).view(self.ngram, -1, real_tgt_len, self.head_dim)], 2)
+
+        # [ngram, B*head, T, 2*T]
+        attn_weights_ngram = torch.einsum('nbtc,nbsc->nbts', (q_ngram, k_ngram))
+
+        # [ngram, B*head, T, S]
+        predict_relative_logits = self.ngram_relative_logits(h_ngram, attn_weights_ngram, real_positions,
+                                                             i_bucket_relative_stream)
+        # [ngram, B*head, T, 2*T]
+        attn_weights_ngram = attn_weights_ngram + predict_relative_logits
+
+        if ngram_mask_matrix is not None:
+            ngram_mask_matrix = ngram_mask_matrix.unsqueeze(1)
+            attn_weights_ngram = attn_weights_ngram + ngram_mask_matrix
+
+        attn_weights_ngram = softmax(
+            attn_weights_ngram, dim=-1, onnx_trace=self.onnx_trace,
+        ).type_as(attn_weights_ngram)
+        attn_weights_ngram = F.dropout(attn_weights_ngram, p=self.dropout, training=self.training)
+
+        # [ngram, B*head, T, c]
+        attn_ngram = torch.einsum('nbts,nbsc->nbtc', (attn_weights_ngram, v_ngram))
+        # [ngram, T, B, C]
+        attn_ngram = attn_ngram.transpose(1, 2).contiguous().view(self.ngram, real_tgt_len, bsz, embed_dim)
+        attn_ngram = self.out_proj(attn_ngram)
+
+        attn_result = []
+        attn_result.append(attn_main)
+        attn_result.append(attn_ngram)
+
+        # [1+ngram*T, B, C]
+        attn = torch.cat(attn_result, 0).view(-1, bsz, embed_dim)
```

Comment:
would be simpler and more readable to just do `attn = torch.cat([attn_main, attn_ngram], 0).view(-1, bsz, embed_dim)` and delete 3 lines above IMO.